### PR TITLE
Refactor and generalize distributed actor transport SIL generation.

### PR DESCRIPTION
**Pull request by DougGregor**
Refactor the code that generates SIL to call into the distributed actor
transport to eliminate duplication and better cope with concrete actor
transports. Centralize the knowledge of which actor transport is used
with a given distributed actor type.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

File: include/swift/SILOptimizer/Utils/DistributedActor.h [hunk 1]
```diff
@@ -13,21 +13,43 @@
...
+/// provided as one of the arguments.
+/// \param args The arguments provided to the call, not including the transport.
+/// \param tryTargets For a call that can throw, the normal and error basic
+/// blocks that the call will branch to.
+void emitActorTransportWitnessCall(
+    SILBuilder &B, SILLocation loc, DeclName methodName,
+    SILValue transport, SILType actorType, llvm::ArrayRef<SILValue> args,
+    llvm::Optional<std::pair<SILBasicBlock *, SILBasicBlock *>> tryTargets =
+        llvm::None);
+
```
Comment (ktoso) [inline include/swift/SILOptimizer/Utils/DistributedActor.h:52]: Oh I see, awesome thanks!

File: lib/SILGen/SILGenDistributed.cpp [hunk 2]
```diff
@@ -52,12 +52,15 @@ static VarDecl* lookupProperty(NominalTypeDecl *ty, DeclName name) {
...
-                   IsNotTake, IsInitialization);
+  CanType propType = SGF.F.mapTypeIntoContext(prop->getInterfaceType())
+      ->getCanonicalType();
+  SILType loweredPropType = SGF.getLoweredType(propType);
+  auto fieldAddr = SGF.B.createRefElementAddr(loc, actorSelf, prop, loweredPropType);
+
+  if (fieldAddr->getType().isAddress())
+    SGF.B.createCopyAddr(loc, value, fieldAddr, IsNotTake, IsInitialization);
+  else
+    SGF.B.emitStoreValueOperation(loc, value, fieldAddr, StoreOwnershipQualifier::Init);
```
Comment (ktoso) [inline lib/SILGen/SILGenDistributed.cpp:63]: Ah, I finally started to understand those differences somewhat now... thanks!

File: lib/SILGen/SILGenDistributed.cpp [hunk 3]
```diff
@@ -161,85 +137,27 @@ static void emitIdentityInit(SILGenFunction &SGF, ConstructorDecl *ctor,
...
   auto temp = SGF.emitTemporaryAllocation(loc, resultTy);
 
-  // ---- actually call transport.assignIdentity(Self.self)
-  B.createApply(
-      loc, assignWitnessMethod, subs,
-      { temp, selfMetatypeValue, transportValue});
+  // --- emit the call itself.
+  emitActorTransportWitnessCall(
+      B, loc, C.Id_assignIdentity, transport, SGF.getLoweredType(selfTy),
+      { temp, selfMetatypeValue });
```
Comment (ktoso) [inline lib/SILGen/SILGenDistributed.cpp:158]: <3

File: lib/SILGen/SILGenDistributed.cpp [hunk 4]
```diff
@@ -282,74 +200,15 @@ static void createDistributedActorFactory_resolve(
...
-  SmallVector<SILValue, 3> params;
-  params.push_back(identityValue);
-  params.push_back(selfMetatypeValue);
-  params.push_back(transportArchetypeValue); // self for the call, as last param
-
-  B.createTryApply(loc, resolveWitnessMethod, subs, params, normalBB, errorBB);
+  emitActorTransportWitnessCall(
+      B, loc, C.Id_resolve, transportValue, SGF.getLoweredType(selfTy),
+      { identityValue, selfMetatypeValue },
+      std::make_pair(normalBB, errorBB));
```
Comment (ktoso) [inline lib/SILGen/SILGenDistributed.cpp:211]: ❤️  ! So much cleaner, thank you!

File: include/swift/AST/KnownSDKTypes.def [hunk 5]
```diff
@@ -45,7 +45,6 @@ KNOWN_SDK_TYPE_DECL(Concurrency, TaskLocal, ClassDecl, 1)
 
 // Distributed actors
 KNOWN_SDK_TYPE_DECL(Distributed, DistributedActor, ProtocolDecl, 0)
-KNOWN_SDK_TYPE_DECL(Distributed, ActorTransport, ProtocolDecl, 0)
```
Comment (ktoso) [inline include/swift/AST/KnownSDKTypes.def:48]: Oh I see... we just need it from `KnownProtocols.def`... thanks!

**Review by ktoso — COMMENTED** [review]
Late LGTM, nothing really to comment other a lot of "wow this is great" 😉 
Thanks, learnt a lot from this PR!

**Pull request merged**